### ISSUE TITLE
Polish Workflow Refactoring API

### ISSUE THREAD

**Issue by jmchilton** [galaxyproject/galaxy#11005]
Post merge of #10988 which introduced this API.

- [ ] Add new refactoring steps:
  - [x] ``{action: 'upgrade_subworkflow', step <StepReference>}`` https://github.com/galaxyproject/galaxy/pull/11032
  - [x] ``{action: 'upgrade_tool', step <StepReference>}`` https://github.com/galaxyproject/galaxy/pull/11032
  - [x] ``{action: 'fill_step_defaults', step: <StepReference>}`` https://github.com/galaxyproject/galaxy/pull/11032
  - [x] ``{action: 'fill_defaults'}`` https://github.com/galaxyproject/galaxy/pull/11032
  - [x] ``{action: 'update_output_label'}``  https://github.com/galaxyproject/galaxy/pull/11032 (not important, but helped with testing)
  - [ ] ``{action: 'restore_workflow_version'}``
- [x] Rework response to include a structured message for each input action. So things like disconnected inputs/outputs during subworkflow upgrades of tool state upgrades can be presented to the user and tracked. https://github.com/galaxyproject/galaxy/pull/11032
- [x] In ``dry_run=true``, capture the messages and do some extra checking that could predict database problems (specifically, I'm sure we could catch label problems at this step and we probably don't.))  https://github.com/galaxyproject/galaxy/pull/11032
- [x] ``dry_run=true`` should return the effective new state, so the UI can do stuff like ``upgrade_subworkflow`` or ``restore_workflow_version``, render the results and display the messages but still require a "Save" before the new version is committed.  https://github.com/galaxyproject/galaxy/pull/11032
- [ ] Store typed and structured actions and messages in the database along with the new workflow version created.
- [ ] Rework ``hasChanges`` tracking in ``Index.vue`` to track structured list of changes, use the refactoring API when possible.
- [ ] Eventually use refactoring API for all UI interactions that touch the ``Workflow`` object (something like editing the tags might be fine with the old API). (This would essentially implement #9166).

**Comment by mvdbeek** [comment]
Do you think it'd be possible to also extend this to check for invalid references ? E.g referring to #{some_output}, when `some_output` is not a tool output?

**Comment by jmchilton** [comment]
> Do you think it'd be possible to also extend this to check for invalid references ? E.g referring to #{some_output}, when some_output is not a tool output?

Tracked that on https://github.com/galaxyproject/galaxy/issues/10989 - seems like a really good idea. I think the client has everything that would be needed - I think recovering the logic for how we reference nested inputs on the client would be the hardest part?

**Comment by simonbray** [comment]
What about a `UpgradeWorkflow` action to update all tools and subworkflows automatically? I think that would be useful both through the API and the UI. Currently, as far as I understand, you have to loop through all workflow steps with `UpgradeTool` and `UpgradeSubworkflow` actions.

**Comment by jmchilton** [comment]
UpgradeWorkflow seems like a solid idea to me! I don't know if I'll have time to work on it in Q1 but it is a good thing to keep in mind. I think the hardest part would simply be fixing up the tool upgrade action to the point that we would want to expose it to users. The subworkflow upgrade logic for dropping connections and workflow outputs would need to be refactored to be reused when apply the upgrade tool action. After that was working, I think it would simply be a matter of looping over the workflow and executing these steps when needed.

**Comment by simonbray** [comment]
Maybe I could work on this?

> the hardest part would simply be fixing up the tool upgrade action to the point that we would want to expose it to users

Can you elaborate? I guess you are referring to this comment here: https://github.com/galaxyproject/galaxy/blob/ca8435681ae36ebb03717b78c76c7e52089e6b10/lib/galaxy/workflow/refactor/execute.py#L484

**Comment by jmchilton** [comment]
Yeah - that comment. There is a bunch of code in the prior method that works on subworkflow modules that can be refactored and reused for tools.
